### PR TITLE
Upload first version of Zoowsome (Assignment 3 + Twist 2)

### PR DESCRIPTION
**Pull request by gellert-kovacs-qubiz**
I know I am several weeks late but I hope at least what I did is correct. :)

I have some problems understanding how this Factory and Abstract Factory design pattern works so I'll have to read more about it.

In the mean time a simple question: 
In the MainController a1 is an Animal but when I cast it to Cow (c1) normalBodyTemp and percBodyHair take the random values which were generated in the factory and not the default values defined in the empty constructors. This is a really nice behavior from my perspective but I don't understand where a1 keeps these two numbers until it is cast to Cow? (I hope you understand what I mean...)

Anyway, thank you for looking through my assignment in advance!

### CODE REVIEW

File: Students/Kovacs Gellert/Zoowsome/src/javasmmr/zoowsome/controllers/MainController.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import javasmmr.zoowsome.models.animals.*;
+import javasmmr.zoowsome.models.animals.Mammal;
+import javasmmr.zoowsome.services.factories.AnimalFactory;
+import javasmmr.zoowsome.services.factories.Constants;
+import javasmmr.zoowsome.services.factories.MammalFactory;
+import javasmmr.zoowsome.services.factories.SpeciesFactory;
+
+public class MainController {
+	public static void main(String[] args) throws Exception {
+		AnimalFactory abstractFactory = new AnimalFactory();
```
Comment (Alex-Cosma) [inline Students/Kovacs Gellert/Zoowsome/src/javasmmr/zoowsome/controllers/MainController.java]: An excellent question you got there. The fact of the matter is that you are creating a `Cow` from the get go. You are not creating only an animal. Even though you are writing
`Animal a1 = new Cow()`
you are actually creating a new `Cow`! However, you are only able to use the methods and attributes from the animal class. 

This is known as polymorphism. 
Imagine the following example:

`Musician musician = new Pianist();`

or

`Musician musician = new Drummer();`

Suppose `Musician` class has a method: `play()`. It doesn't matter who the musician is, if you call play method, you'll know how to play the determined instrument, base on concrete class, on this case, no matter if Pianist or Drummer.

Each concrete class, overwriting its `play()` method, can play on your own:

e.g. for `Pianist` class: `play() { playsBeethoven(); }`
or for `Drummer` class: `play() { playsDreamTheater(); }`

For detailed information, please check http://docs.oracle.com/javase/tutorial/java/IandI/polymorphism.html

It's always good to remember to use it with inheritance http://docs.oracle.com/javase/tutorial/java/IandI/subclasses.html

**Comment by Alex-Cosma** [comment]
Which part of the factory methods do you not understand? Are you sure you understand the purpose of using such factories? If not, maybe that's the thing which boggles you, not the actual implementation.

**Comment by gellert-kovacs-qubiz** [comment]
Thank you for your answer! Related to the factory methods as far as I can see they are used to hide the creation process of animals from the user but I'm not sure. You are right, I don't really understand why we need them, the actual implementation is simple.

**Comment by Alex-Cosma** [comment]
If by `user` you mean the programmer, yes. If not, then I don't fully understand which _user_ you are talking about

**Comment by gellert-kovacs-qubiz** [comment]
That is exactly what I meant. Thank you.

**Comment by Alex-Cosma** [comment]
Good work

**Pull request merged**